### PR TITLE
8267191: Avoid repeated SystemDictionaryShared::should_be_excluded calls

### PR DESCRIPTION
**Pull request by iklam**
`SystemDictionaryShared::should_be_excluded` is called during CDS dump time to see if a class should be excluded from the archive. This function calls itself to check super types. As a result, some classes many be checked more than once.

The repeated calls slows down dumping. Also, If a class is excluded, we will see multiple warning messages like this, which is too verbose and confusing.

```
Skipping org/eclipse/osgi/util/NLS: Signed JAR
Skipping org/eclipse/osgi/util/NLS: Signed JAR
Skipping org/eclipse/osgi/util/NLS: Signed JAR
```

BTW, to make the code a little easier to read, I also changed

```
 warn_excluded(k, "Failed verification");
 return true;
```

to

```
return  warn_excluded(k, "Failed verification");
```

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8267191](https://bugs.openjdk.java.net/browse/JDK-8267191): Avoid repeated SystemDictionaryShared::should_be_excluded calls

### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to b31408a451338e05a54ccbad883ea63f9bfa7911
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to f1e1e276fc00ef79d5c0f09e551b116e959b050a

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4070/head:pull/4070` \
`$ git checkout pull/4070`

Update a local copy of the PR: \
`$ git checkout pull/4070` \
`$ git pull https://git.openjdk.java.net/jdk pull/4070/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4070`

View PR using the GUI difftool: \
`$ git pr show -t 4070`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4070.diff">https://git.openjdk.java.net/jdk/pull/4070.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/classfile/systemDictionaryShared.cpp [hunk 1]
```diff
@@ -1393,34 +1401,33 @@ bool SystemDictionaryShared::should_be_excluded(InstanceKlass* k) {
...
-    return true;
+    return warn_excluded(k, "Old class has been linked");
   }
 
-  InstanceKlass* super = k->java_super();
-  if (super != NULL && should_be_excluded(super)) {
+  if (k->is_hidden() && !is_registered_lambda_proxy_class(k)) {
     ResourceMark rm;
-    log_warning(cds)("Skipping %s: super class %s is excluded", k->name()->as_C_string(), super->name()->as_C_string());
+    log_debug(cds)("Skipping %s: %s", k->name()->as_C_string(), "Hidden class");
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/systemDictionaryShared.cpp:1409]: You don't need to feed "Hidden class" to a %s.

File: src/hotspot/share/classfile/systemDictionaryShared.cpp [hunk 2]
```diff
@@ -1393,34 +1401,33 @@ bool SystemDictionaryShared::should_be_excluded(InstanceKlass* k) {
...
     return true;
   }
 
-  if (k->is_hidden() && !is_registered_lambda_proxy_class(k)) {
+  InstanceKlass* super = k->java_super();
+  if (super != NULL && check_for_exclusion(super, NULL)) {
     ResourceMark rm;
-    log_debug(cds)("Skipping %s: %s", k->name()->as_C_string(), "Hidden class");
+    log_warning(cds)("Skipping %s: super class %s is excluded @%p", k->name()->as_C_string(), super->name()->as_C_string(), super);
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/systemDictionaryShared.cpp:1416]: Pretty sure we still don't/can't use %p in shared code, but should be INTPTR_FORMAT and use p2i(super)

File: src/hotspot/share/classfile/systemDictionaryShared.cpp [hunk 3]
```diff
@@ -1393,34 +1401,33 @@ bool SystemDictionaryShared::should_be_excluded(InstanceKlass* k) {
...
   }
 
   Array<InstanceKlass*>* interfaces = k->local_interfaces();
   int len = interfaces->length();
   for (int i = 0; i < len; i++) {
     InstanceKlass* intf = interfaces->at(i);
-    if (should_be_excluded(intf)) {
-      log_warning(cds)("Skipping %s: interface %s is excluded", k->name()->as_C_string(), intf->name()->as_C_string());
+    if (check_for_exclusion(intf, NULL)) {
+      log_warning(cds)("Skipping %s: interface %s is excluded @%p", k->name()->as_C_string(), intf->name()->as_C_string(), intf);
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/systemDictionaryShared.cpp:1425]: Ditto on %p

**Review by dholmes-ora — APPROVED** [review]
Hi Ioi,

These changes seem fine.

A couple of nits below.

Thanks,
David

**Review by coleenp — APPROVED** [review]
Looks good!

File: src/hotspot/share/classfile/systemDictionaryShared.cpp (hunk 2 above)
Comment (iklam) [reply src/hotspot/share/classfile/systemDictionaryShared.cpp]: I reverted back to the original code and removed the %p. I added %p for temporary debugging only.

File: src/hotspot/share/classfile/systemDictionaryShared.cpp (hunk 1 above)
Comment (iklam) [reply src/hotspot/share/classfile/systemDictionaryShared.cpp]: Fixed.

**Review by dholmes-ora — APPROVED** [review]
LGTM.

**Comment by iklam** [comment]
Thanks @coleenp and @dholmes-ora for the review!
/integrate

**Pull request closed without merging**